src/test/plugins/ucc_discovery_plugin_test.cpp

```diff
@@ -277,19 +490,108 @@ TEST_P(UccDiscoveryPluginMultiEncodingTest, PluginFullRun) {
+  EXPECT_EQ(predicate_column_A_constraint->last_validation_result(), ValidationResultType::VALID);
 
-  EXPECT_TRUE(constraints_B.contains({{_join_columnB->original_column_id}, KeyConstraintType::UNIQUE}));
-  EXPECT_FALSE(constraints_B.contains({{_predicate_column_B->original_column_id}, KeyConstraintType::UNIQUE}));
+  const auto join_column_B_constraint =
+      constraints_B.find(TableKeyConstraint{{_join_columnB->original_column_id}, KeyConstraintType::UNIQUE});
+  ASSERT_NE(join_column_B_constraint, constraints_B.end());
+  EXPECT_EQ(join_column_B_constraint->last_validation_result(), ValidationResultType::VALID);
+
+  const auto predicate_column_B_constraint =
+      constraints_B.find(TableKeyConstraint{{_predicate_column_B->original_column_id}, KeyConstraintType::UNIQUE});
+  ASSERT_NE(predicate_column_B_constraint, constraints_B.end());
+  EXPECT_EQ(predicate_column_B_constraint->last_validation_result(), ValidationResultType::INVALID);
 
   // Ensure we clear the plan caches.
   EXPECT_EQ(Hyrise::get().default_lqp_cache->size(), 0);
   EXPECT_EQ(Hyrise::get().default_pqp_cache->size(), 0);
 }
 
+TEST_F(UccDiscoveryPluginMultiEncodingTest, PluginIntegrationTestWithInvalidation) {
+  Hyrise::get().default_pqp_cache = std::make_shared<SQLPhysicalPlanCache>();
+  Hyrise::get().default_lqp_cache = std::make_shared<SQLLogicalPlanCache>();
+
+  const auto exec_groupby_statement = [&]() {
+    auto pipeline = SQLPipelineBuilder{"SELECT b,c FROM " + _table_name_A + " GROUP BY b,c;"}.create_pipeline();
+    const auto& [pipeline_status, _] = pipeline.get_result_table();
+    EXPECT_EQ(pipeline_status, SQLPipelineStatus::Success);
+  };
+
+  const auto exec_insert_statement = [&]() {
+    auto transaction_context = Hyrise::get().transaction_manager.new_transaction_context(AutoCommit::No);
+    auto pipeline = SQLPipelineBuilder{"INSERT INTO " + _table_name_A + " (A_a, b, c) VALUES (14, 12, 'duplicate_b');"}
+                        .with_transaction_context(transaction_context)
+                        .create_pipeline();
+    const auto& [pipeline_status, _] = pipeline.get_result_table();
+    EXPECT_EQ(pipeline_status, SQLPipelineStatus::Success);
+    transaction_context->commit();
+  };
+
+  // Perform a single query to populate the LQPCache for `identify_ucc_candidates`.
+  exec_groupby_statement();
+
+  EXPECT_EQ(Hyrise::get().default_lqp_cache->size(), 1);
+
+  // Discover UCCs used in previous query.
+  auto& plugin_manager = Hyrise::get().plugin_manager;
+  plugin_manager.load_plugin(build_dylib_path("libhyriseUccDiscoveryPlugin"));
+  plugin_manager.exec_user_function("hyriseUccDiscoveryPlugin", "DiscoverUCCs");
+
+  // Check that key constraint was added to the table.
+  const auto constraint_A_c = TableKeyConstraint{{ColumnID{2}}, KeyConstraintType::UNIQUE};
+
+  EXPECT_EQ(_table_A->soft_key_constraints().size(), 2);
+  EXPECT_EQ(_table_A->soft_key_constraints().find(constraint_A_c)->last_validation_result(),
```

Comment:
This can segfault if the constraint was not set, right? Please Split into two checks as you did below.